provider/oci/images_test.go

```diff
@@ -0,0 +1,393 @@
+	c.Check(imgType.Id, gc.Equals, i.testImageID)
+}
+
+func (i *imagesSuite) TestNewInstanceImageUnknownOS(c *gc.C) {
+	image := ociCore.Image{
+		CompartmentId:          &i.testCompartment,
+		Id:                     &i.testImageID,
+		OperatingSystem:        makeStringPointer("NotKnownToJuju"),
+		OperatingSystemVersion: makeStringPointer("14.04"),
+		DisplayName:            makeStringPointer("Canonical-Ubuntu-14.04-2018.01.11-0"),
+	}
+
+	_, err := oci.NewInstanceImage(image, &i.testCompartment)
+	c.Assert(err, gc.ErrorMatches, "os NotKnownToJuju not supported")
+}
+
+func (i *imagesSuite) TestRefreshImageCache(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	compute := ocitesting.NewMockOCIComputeClient(ctrl)
+	defer ctrl.Finish()
+
+	fakeUbuntuIDSecond := "fakeUbuntu2"
+
+	listImageRequest, listImageResponse := makeListImageRequestResponse(
+		i.testCompartment, []map[string]string{
+			map[string]string{
+				"id":      "fakeUbuntu1",
+				"os":      "Canonical Ubuntu",
+				"name":    "Canonical-Ubuntu-14.04-2018.01.11-0",
+				"version": "14.04",
+			},
+			map[string]string{
+				"id":      fakeUbuntuIDSecond,
+				"os":      "Canonical Ubuntu",
+				"name":    "Canonical-Ubuntu-14.04-2018.01.12-0",
+				"version": "14.04",
+			},
+			map[string]string{
+				"id":      "fakeCentOS",
+				"os":      "CentOS",
+				"name":    "CentOS-7-2017.10.19-0",
+				"version": "7",
+			},
+		})
+
+	shapesRequestUbuntu, shapesResponseUbuntu := makeShapesRequestResponse(
+		i.testCompartment, "fakeUbuntu1", []string{"VM.Standard2.1", "VM.Standard1.2"})
+
+	shapesRequestUbuntuSecond := ociCore.ListShapesRequest{
+		CompartmentId: &i.testCompartment,
+		ImageId:       &fakeUbuntuIDSecond,
+	}
+
+	shapesRequestCentOS, shapesResponseCentOS := makeShapesRequestResponse(
+		i.testCompartment, "fakeCentOS", []string{"VM.Standard1.2"})
+
+	gomock.InOrder(
+		compute.EXPECT().ListImages(context.Background(), listImageRequest).Return(listImageResponse, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestUbuntu).Return(shapesResponseUbuntu, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestUbuntuSecond).Return(shapesResponseUbuntu, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestCentOS).Return(shapesResponseCentOS, nil),
+	)
+
+	imgCache, err := oci.RefreshImageCache(compute, &i.testCompartment)
+	c.Assert(err, gc.IsNil)
+	c.Assert(imgCache, gc.NotNil)
+	c.Check(len(imgCache.ImageMap()), gc.Equals, 2)
+
+	imageMap := imgCache.ImageMap()
+	c.Check(len(imageMap["trusty"]), gc.Equals, 2)
+	c.Check(len(imageMap["centos7"]), gc.Equals, 1)
+
+	timeStamp, _ := time.Parse("2006.01.02", "2018.01.12")
+
+	// Check that the first image in the array is the newest one
+	c.Assert(imageMap["trusty"][0].Version.TimeStamp, gc.Equals, timeStamp)
+
+	// Check that InstanceTypes are set
+	c.Assert(len(imageMap["trusty"][0].InstanceTypes), gc.Equals, 2)
+	c.Assert(len(imageMap["centos7"][0].InstanceTypes), gc.Equals, 1)
```

Comment:
I was a bit confused by this at first, because len(imageMap["trusty"]) is 2, but here you are checking the VM.Standard.1.1 etc.
I think this would be a case where it would be a lot clearer to use 'DeepEquals':
c.Check(imageMap["trusty"][0].InstanceTypes, gc.DeepEquals, []string{"VM.Standard1.2", ...})

Reply:
```InstanceTypes``` actually contains a ```[]instances.InstanceType```. I could compose that object and compare if you prefer that, but it is a bit more involved than a string array.